pkg/provisioner/crdprovisioner.go

```diff
@@ -0,0 +1,320 @@
+			NodeName:         nodeID,
+			VolumeContext:    volumeContext,
+			RequestedRole:    v1alpha1.PrimaryRole,
+		},
+	}
+
+	attachmentCreated, err := azVA.Create(ctx, azVolumeAttachment, metav1.CreateOptions{})
+	if err != nil {
+		klog.Error("Error creating azvolume attachment for volume id (%s) to node id (%s) error : %v", volumeID, nodeID, err)
+		return nil, err
+	}
+
+	conditionFunc := func() (bool, error) {
+		if attachmentCreated.Status == nil {
+			return false, status.Error(codes.Internal, fmt.Sprintf("Failed to attach azvolume attachment resource for volume id (%s) to node (%s)", volumeID, nodeID))
+		}
+
+		return true, nil
+	}
+
+	err = wait.PollImmediate(interval, timeout, conditionFunc)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return attachmentCreated.Status.PublishContext, nil
+}
+
+func (c *CrdProvisioner) UnpublishVolume(
+	ctx context.Context,
+	volumeID string,
+	nodeID string,
+	secrets map[string]string) error {
+	azVA := c.azDiskClient.DiskV1alpha1().AzVolumeAttachments(c.namespace)
+	err := azVA.Delete(ctx, azureutils.GetAzVolumeAttachmentName(volumeID, nodeID), metav1.DeleteOptions{})
+	if errors.IsNotFound(err) {
+		klog.Info("Could not find the volume attachment (%s). Deletion succedded", volumeID)
+		return nil
+	}
+
+	if err != nil {
+		klog.Error("Failed to delete azvolume attachment resource for volume id (%s) to node (%s), error: %v", volumeID, nodeID, err)
+		return err
+	}
+
+	conditionFunc := func() (bool, error) {
+		// Verify if the azVolume is deleted
+		_, err = azVA.Get(ctx, azureutils.GetAzVolumeAttachmentName(volumeID, nodeID), metav1.GetOptions{})
+		if errors.IsNotFound(err) {
+			return true, nil
+		}
+
+		return false, status.Error(codes.Internal, fmt.Sprintf("Failed to delete azvolume resource attachment for volume name (%s) to node (%s)", volumeID, nodeID))
+	}
+
+	return wait.PollImmediate(interval, timeout, conditionFunc)
+}
+
+func (c *CrdProvisioner) ExpandVolume(
+	ctx context.Context,
+	volumeID string,
+	capacityRange *v1alpha1.CapacityRange,
+	secrets map[string]string) (*v1alpha1.AzVolumeStatusParams, error) {
+	azV := c.azDiskClient.DiskV1alpha1().AzVolumes(c.namespace)
+	azVolume, err := azV.Get(ctx, volumeID, metav1.GetOptions{})
+	if err != nil || azVolume == nil {
+		klog.Errorf("Failed to retrieve existing volume id (%s)", volumeID)
+		return nil, status.Error(codes.Internal, fmt.Sprintf("Failed to retrieve volume id (%s), error: %v", volumeID, err))
+	}
+
+	azVolume.Spec.CapacityRange = capacityRange
+	azVolumeUpdated, err := azV.Update(ctx, azVolume, metav1.UpdateOptions{})
+	if err != nil || azVolumeUpdated == nil {
+		klog.Errorf("Failed to update azvolume resource for volume name (%s), error: %v", volumeID, err)
+		return nil, err
+	}
+
+	if azVolumeUpdated.Status == nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("Unable to fetch status of volume created for volume name (%s)", volumeID))
+	}
+
+	return azVolumeUpdated.Status.ResponseObject, nil
```

Comment:
We need to wait (polling again) for the volume expansion to complete; that is, wait for `azVolumeUpdated.Status.ResponseObject.CapacityRange` to match the requested capacity.